### PR TITLE
MouseMoveEvent now populates the From field on OSX

### LINKED ISSUE #42: OSX - MouseMovedEvent From field is not set
On Mac, calling  ` e.(wde.MouseMovedEvent).From` always yields an empty point with ordinates `(0,0)`.

  [The code](https://github.com/skelterjohn/go.wde/blob/master/cocoa/events_darwin.go) shows that `From` is never set.  I'm not familiar mouse events on Mac so I don't know if that value is provided or if it will need to be stored within this library.

I'd be willing to help work on a fix if I can get a little direction on how this should be done.  It seems if nothing else, `From` can be stored in the Window struct for darwin and updated on each MouseMovedEvent.

Thanks!

### PR DESCRIPTION
**Pull request by ricochet2200**
Fixes #42

### CODE REVIEW

File: cocoa/events_darwin.go [hunk 1]
```diff
@@ -57,33 +62,73 @@ func (w *Window) EventChan() (events <-chan interface{}) {
...
 			case C.GMDMouseMoved:
-				var me wde.MouseMovedEvent
-				me.Where.X = int(e.data[0])
-				me.Where.Y = int(e.data[1])
-				ec <- me
+				var mme wde.MouseMovedEvent
+				mme.Where.X = int(e.data[0])
+				mme.Where.Y = int(e.data[1])
+				if lastX != noX {
+					mme.From.X = int(lastX)
```
Comment (skelterjohn) [inline cocoa/events_darwin.go]: lastX and lastY are already ints

Comment (skelterjohn) [inline cocoa/events_darwin.go]: ah, in the file you copied from they were int32s

File: cocoa/events_darwin.go [hunk 2]
```diff
@@ -46,6 +46,11 @@ func (w *Window) EventChan() (events <-chan interface{}) {
 	downKeys := make(map[string]bool)
 	ec := make(chan interface{})
 	go func(ec chan<- interface{}) {
+
+		var noX int = 1<<31 - 1
```
Comment (skelterjohn) [inline cocoa/events_darwin.go]: int32 if you're going to copy this weirdness with the bit offset. i honestly don't remember how this was arrived at.

Comment (ricochet2200) [inline cocoa/events_darwin.go]: Would it be okay to set noX to `-1`?  `-1` will always be a invalid value whereas `var noX int = 1<<31 - 1` could very well be valid (although presumably very unlikely).

If yes, would you also like me to update it on the linux side?

Comment (ricochet2200) [inline cocoa/events_darwin.go]: Actually, I'm not in close to a linux computer.  I'll let someone else change the linux code if that is desired.

File: cocoa/events_darwin.go [hunk 3]
```diff
@@ -46,6 +47,10 @@ func (w *Window) EventChan() (events <-chan interface{}) {
 	downKeys := make(map[string]bool)
 	ec := make(chan interface{})
 	go func(ec chan<- interface{}) {
+
+		noX := -1
+		lastXY := image.Point{-1, -1}
```
Comment (skelterjohn) [inline cocoa/events_darwin.go]: var lastXY image.Point
lastXY.X = noX

Comment (skelterjohn) [inline cocoa/events_darwin.go]: Or, 
var noPoint = image.Point{-1, -1}
var lastXY = noPoint

and compare lastXY to noPoint instead of lastXY.X to noX

Comment (ricochet2200) [inline cocoa/events_darwin.go]: hmmm... this technically works as is but is crap.  I was in a rush to fix it and should have paid better attention.  Fix coming soon.

**Comment by sqweek** [comment]
Rick: I've merged your fix into a branch including several other changes, is there any chance you can give it a quick test before I update master?

I only have a dodgy OSX VM to test in, so I'd like some reassurance that my changes aren't causing problems on the real thing :) I've pushed to branch 'sqweek' for now.

**Comment by ricochet2200** [comment]
@sqweek Is there anyway in particular you'd like it to be tested?  My quick test is to change the imports from `github.com/skelterjohn/<whatever>` to a local import (ie. `"../cocoa"`) and the `go run wdetest.go`.

Is there a {better,different} way you'd like me to do?

**Comment by sqweek** [comment]
@ricochet2200 wdetest is perfect. I'm not sure if just changing the imports in `wdetest.go` is sufficient, as the modules you're importing may still refer to `github.com/skelterjohn` -- but if you changed all imports in your clone that should do it. For this kind of thing I would tend to cd into GOPATH/GOROOT (wherever go has downloaded the package) and then `cd src/github.com/skelterjohn/go.wde && git fetch && git checkout origin/sqweek`. That way you don't have to change imports, just have to remember to switch back to origin/master after testing :)

The other thing added on the branch is a mouse cursor API - if you press 1 2 or 3 in wdetest it should switch between different mouse cursors (but only for the window with focus).

**Comment by skelterjohn** [comment]
On Mon, Aug 17, 2015 at 9:06 PM, sqweek <notifications@github.com> wrote:

>  and then cd src/github.com/skelterjohn/go.wde && git fetch && git
> checkout origin/sqweek. That way you don't have to change imports, just
> have to remember to switch back to origin/master after testing
>
big +1 on this. that is the appropriate way to work on someone else's go
git repo.

**Comment by ricochet2200** [comment]
I was pretty confident I was doing it wrong, but wasn't sure the right way to do it.  Other projects I have worked with just use relative paths.  I think at one point it was considered an undocumented "feature", but I'm not sure if that is still true.

I also didn't realize GOROOT stored the src as a git repository, I thought it was just an export...the repo definitely makes it more powerful.  

@sqweek I followed your instructions exactly and when I run the test I get this error:

    $GOPATH/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:96: undefined: wde.Cursor

I must be importing wrong, I'll have to think about it some more.  I've got to run for now, but I'll have more time to look at it later today.

**Comment by sqweek** [comment]
Hrm I don't know how that error would come about. If cocoa/cocoa_darwin.go is referencing `wde.Cursor` you must be on my branch, but then you should also have cursor.go in the top-level dir which defines the Cursor type. it does run in my VM:

```
$ go version
go version go1.4.2 darwin/amd64
$ git status
HEAD detached at origin/sqweek
nothing to commit, working directory clean
$ go run wdetest/wdetest.go
resize 200 200
resize 200 200
```

Thanks for your effort so far, if it really doesn't want to work then I'll push your fix separately and try and rope someone else into OSX testing.

**Comment by skelterjohn** [comment]
I can try to find time to test tonight. I have a mac :)

**Comment by ricochet2200** [comment]
Thanks for posting `go version`, apparently I was still running go 1.3 so I upgraded.  I upgraded to 1.4 and it ran fine...that seems a little weird.  I was in rush this morning so maybe I just made a mistake somewhere.

    go.wde :) go version
    go version go1.4.2 darwin/amd64

    $ git status
    HEAD detached at origin/sqweek
    nothing to commit, working directory clean

However when I ran it again it crashed:




fatal error: unexpected signal during runtime execution
[signal 0xb code=0x1 addr=0x19 pc=0x7fff8a28a79b]

runtime stack:
runtime.gothrow(0x41537f0, 0x2a)
	/usr/local/go/src/runtime/panic.go:503 +0x8e
runtime.sigpanic()
	/usr/local/go/src/runtime/sigpanic_unix.go:14 +0x5e

goroutine 6 [syscall, locked to thread]:
runtime.cgocall_errno(0x40011f0, 0xc208019590, 0x0)
	/usr/local/go/src/runtime/cgocall.go:130 +0xf5 fp=0xc208019570 sp=0xc208019548
github.com/skelterjohn/go.wde/cocoa._Cfunc_openWindow(0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/:98 +0x4d fp=0xc208019590 sp=0xc208019570
github.com/skelterjohn/go.wde/cocoa.NewWindow(0xc8, 0xc8, 0x41da260, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:101 +0x31 fp=0xc2080195c0 sp=0xc208019590
github.com/skelterjohn/go.wde/cocoa.func·001(0xc8, 0xc8, 0x0, 0x0, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:58 +0x69 fp=0xc2080195f8 sp=0xc2080195c0
github.com/skelterjohn/go%2ewde.NewWindow(0xc8, 0xc8, 0x0, 0x0, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wde.go:110 +0x5d fp=0xc208019630 sp=0xc2080195f8
main.func·002()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:45 +0x8d fp=0xc2080197e0 sp=0xc208019630
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2232 +0x1 fp=0xc2080197e8 sp=0xc2080197e0
created by main.wdetest
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:151 +0xf2

goroutine 1 [syscall, locked to thread]:
github.com/skelterjohn/go.wde/cocoa._Cfunc_NSAppRun()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/:32 +0x45
github.com/skelterjohn/go.wde/cocoa.Run()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:211 +0x1b
github.com/skelterjohn/go%2ewde.Run()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wde.go:87 +0x22
main.main()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:32 +0x30

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2232 +0x1

goroutine 5 [semacquire]:
sync.(*WaitGroup).Wait(0xc20801e060)
	/usr/local/go/src/sync/waitgroup.go:132 +0x169
main.wdetest()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:155 +0x128
created by main.main
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:31 +0x29

goroutine 7 [syscall, locked to thread]:
github.com/skelterjohn/go.wde/cocoa._Cfunc_openWindow(0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/:98 +0x4d
github.com/skelterjohn/go.wde/cocoa.NewWindow(0xc8, 0xc8, 0x405ae0d, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:101 +0x31
github.com/skelterjohn/go.wde/cocoa.func·001(0xc8, 0xc8, 0x0, 0x0, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:58 +0x69
github.com/skelterjohn/go%2ewde.NewWindow(0xc8, 0xc8, 0x0, 0x0, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wde.go:110 +0x5d
main.func·002()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:45 +0x8d
created by main.wdetest
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:153 +0x118
exit status 2


Then I ran it again and it worked.  Then I ran it again and it crashed with a different error: 

wdetest(56262,0xb03a6000) malloc: *** error for object 0x4320ca8: incorrect checksum for freed object - object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug
SIGABRT: abort
PC=0x7fff898c0866
signal arrived during cgo execution

goroutine 7 [syscall, locked to thread]:
runtime.cgocall_errno(0x40011f0, 0xc208019d90, 0x0)
	/usr/local/go/src/runtime/cgocall.go:130 +0xf5 fp=0xc208019d70 sp=0xc208019d48
github.com/skelterjohn/go.wde/cocoa._Cfunc_openWindow(0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/:98 +0x4d fp=0xc208019d90 sp=0xc208019d70
github.com/skelterjohn/go.wde/cocoa.NewWindow(0xc8, 0xc8, 0x405ae0d, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:101 +0x31 fp=0xc208019dc0 sp=0xc208019d90
github.com/skelterjohn/go.wde/cocoa.func·001(0xc8, 0xc8, 0x0, 0x0, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:58 +0x69 fp=0xc208019df8 sp=0xc208019dc0
github.com/skelterjohn/go%2ewde.NewWindow(0xc8, 0xc8, 0x0, 0x0, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wde.go:110 +0x5d fp=0xc208019e30 sp=0xc208019df8
main.func·002()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:45 +0x8d fp=0xc208019fe0 sp=0xc208019e30
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2232 +0x1 fp=0xc208019fe8 sp=0xc208019fe0
created by main.wdetest
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:153 +0x118

goroutine 1 [syscall, locked to thread]:
github.com/skelterjohn/go.wde/cocoa._Cfunc_NSAppRun()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/:32 +0x45
github.com/skelterjohn/go.wde/cocoa.Run()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:211 +0x1b
github.com/skelterjohn/go%2ewde.Run()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wde.go:87 +0x22
main.main()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:32 +0x30

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2232 +0x1

goroutine 5 [semacquire]:
sync.(*WaitGroup).Wait(0xc20801e060)
	/usr/local/go/src/sync/waitgroup.go:132 +0x169
main.wdetest()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:155 +0x128
created by main.main
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:31 +0x29

goroutine 6 [syscall, locked to thread]:
github.com/skelterjohn/go.wde/cocoa._Cfunc_openWindow(0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/:98 +0x4d
github.com/skelterjohn/go.wde/cocoa.NewWindow(0xc8, 0xc8, 0x41da260, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:101 +0x31
github.com/skelterjohn/go.wde/cocoa.func·001(0xc8, 0xc8, 0x0, 0x0, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:58 +0x69
github.com/skelterjohn/go%2ewde.NewWindow(0xc8, 0xc8, 0x0, 0x0, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wde.go:110 +0x5d
main.func·002()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:45 +0x8d
created by main.wdetest
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:151 +0xf2

rax     0x0
rbx     0xb03a6000
rcx     0xb03a57e8
rdx     0x0
rdi     0x5613
rsi     0x6
rbp     0xb03a5810
rsp     0xb03a57e8
r8      0x10
r9      0xfffffff0
r10     0x8000000
r11     0x206
r12     0x421b000
r13     0x45e6000
r14     0x6
r15     0x0
rip     0x7fff898c0866
rflags  0x206
cs      0x7
fs      0x0
gs      0x45e0000

Then I ran it again and it worked.  Running it several times reveals it failing about 50% of the time.  For me, adding the `-race` flag yielded no warnings, but did decrease the frequency of the failures to around 10%.

    go run -race wdetest/wdetest.go

Thoughts?

**Comment by sqweek** [comment]
Well... balls. I notice both stack dumps have two threads in `openWindow` so I wonder if something in there is not threadsafe depending on the OS X/Cocoa version. Does it work if you remove the `[NSApp activateIgnoringOtherApps:YES]` line from `openWindow` (in `cocoa/gmd.m`)?

I'm running OSX 10.9 (Yosemite), with this Cocoa framework:

```
$ plutil -p /System/Library/Frameworks/Cocoa.framework/Versions/A/Resources/version.plist
{
  "CFBundleShortVersionString" => "6.8"
  "BuildVersion" => "269"
  "ProjectName" => "Cocoa"
  "SourceVersion" => "21000000000000"
  "CFBundleVersion" => "21"
}
```

**Comment by ricochet2200** [comment]
I'm still running Maverick:

    plutil -p /System/Library/Frameworks/Cocoa.framework/Versions/A/Resources/version.plist
    {
      "CFBundleShortVersionString" => "6.8"
      "BuildVersion" => "666"
      "ProjectName" => "Cocoa"
      "SourceVersion" => "20000000000000"
      "CFBundleVersion" => "20"
    }

I removed the line in `cocoa/gmd.m` and am still getting the same error (see diff below):

cocoa/gmd.m
-    [NSApp activateIgnoringOtherApps:YES];

**Comment by ricochet2200** [comment]
I'm running git bisect to try to track this down:

    git bisect start
    git bisect bad
    git bisect good 025351e6762a154aa38be2f547ed50bcdb5850f0

If I can run it 10 times straight without an error I mark it as good, else bad.  However, it appears this bug predates this branch:

    git co 025351e6762a154aa38be2f547ed50bcdb5850f0

I ran it 25 times and it crashed only the first time: 

go run wdetest/wdetest.go
wdetest(70917,0xb021d000) malloc: *** error for object 0x4744cd8: incorrect checksum for freed object - object was probably modified after being freed.
*** set a breakpoint in malloc_error_break to debug
SIGABRT: abort
PC=0x7fff898c0866
signal arrived during cgo execution

goroutine 6 [syscall, locked to thread]:
runtime.cgocall_errno(0x4000a20, 0xc208019d90, 0x0)
	/usr/local/go/src/runtime/cgocall.go:130 +0xf5 fp=0xc208019d70 sp=0xc208019d48
github.com/skelterjohn/go.wde/cocoa._Cfunc_openWindow(0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/:84 +0x4d fp=0xc208019d90 sp=0xc208019d70
github.com/skelterjohn/go.wde/cocoa.NewWindow(0xc8, 0xc8, 0x41a9520, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:73 +0x31 fp=0xc208019dc0 sp=0xc208019d90
github.com/skelterjohn/go.wde/cocoa.func·001(0xc8, 0xc8, 0x0, 0x0, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:40 +0x69 fp=0xc208019df8 sp=0xc208019dc0
github.com/skelterjohn/go%2ewde.NewWindow(0xc8, 0xc8, 0x0, 0x0, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wde.go:109 +0x5d fp=0xc208019e30 sp=0xc208019df8
main.func·002()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:45 +0x8d fp=0xc208019fe0 sp=0xc208019e30
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2232 +0x1 fp=0xc208019fe8 sp=0xc208019fe0
created by main.wdetest
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:143 +0xf2

goroutine 1 [syscall, locked to thread]:
github.com/skelterjohn/go.wde/cocoa._Cfunc_NSAppRun()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/:30 +0x45
github.com/skelterjohn/go.wde/cocoa.Run()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:177 +0x1b
github.com/skelterjohn/go%2ewde.Run()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wde.go:86 +0x22
main.main()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:32 +0x30

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2232 +0x1

goroutine 5 [semacquire]:
sync.(*WaitGroup).Wait(0xc20801e040)
	/usr/local/go/src/sync/waitgroup.go:132 +0x169
main.wdetest()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:147 +0x128
created by main.main
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:31 +0x29

goroutine 7 [syscall, locked to thread]:
github.com/skelterjohn/go.wde/cocoa._Cfunc_setWindowSize(0x4740210, 0xc8000000c8)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/:108 +0x45
github.com/skelterjohn/go.wde/cocoa.(*Window).SetSize(0xc20801e080, 0xc8, 0xc8)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:94 +0x86
github.com/skelterjohn/go.wde/cocoa.NewWindow(0xc8, 0xc8, 0xc20801e080, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:77 +0xa5
github.com/skelterjohn/go.wde/cocoa.func·001(0xc8, 0xc8, 0x0, 0x0, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:40 +0x69
github.com/skelterjohn/go%2ewde.NewWindow(0xc8, 0xc8, 0x0, 0x0, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wde.go:109 +0x5d
main.func·002()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:45 +0x8d
created by main.wdetest
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:145 +0x118

rax     0x0
rbx     0xb021d000
rcx     0xb021c568
rdx     0x0
rdi     0x2203
rsi     0x6
rbp     0xb021c590
rsp     0xb021c568
r8      0x10
r9      0xfffffff0
r10     0x8000000
r11     0x206
r12     0x41f8000
r13     0x42ff000
r14     0x6
r15     0x0
rip     0x7fff898c0866
rflags  0x206
cs      0x7
fs      0x0
gs      0x45c0000
exit status 2


1) It appears there has been a race condition in the code for a while, but the new changes make it manifest more often (at least on my machine).  
2) `go run -race wdetest/wdetest.go` is not finding them.  Perhaps this means the race condition lies outside of the go code?

**Comment by sqweek** [comment]
Ah, nice detective work! I'm no closer to understanding which code is at fault, but after many runs I have reproduced a similar crash on my OSX. Definitely not as frequent as on your setup, but I do often see graphical glitches with the title bar -- I'm beginning to wonder if thats the same memory corruption manifesting in a different way...

I'm fairly new to OSX development, but there's a couple of issues at play which are likely involved in the crash.

1. cgo's support for objective-c was improved with go 1.3. cookieo9 prepared a patch for go.wde which took advantage of this and removed the need to manually install the gomacdraw framework. I've included his patch in my branch (commit dce4a6e).
2. OSX/Cocoa has some very particular threading requirements, eg. some functions are only safe when called from the application's main thread. This restriction can be awkward to work with from go, and it's possible go.wde needs to be more careful in this area.

**Comment by skelterjohn** [comment]
go.wde respects the main thread restriction, in theory.

**Comment by ricochet2200** [comment]
@sqweek I'm glad to hear you were able to replicate the crash.

To be clear, I can get `wdetest` to crash using code **prior** to commit `dce4a6e`, so the crashes don't originate with changes made in that commit.  However, they may change the timing of some things that make the race conditions manifest more frequently.

I know slim to nothing about OSX/Cocoa, but this line:

    signal arrived during cgo execution

along with the fact that `-race` isn't detecting anything, makes me think the error is not in the go code.

**Comment by sqweek** [comment]
@skelterjohn I'm not sure I'm sold on the theory - there's definitely calls into cocoa happening from outside the main thread. eg. all Window operations happen on whatever thread the calling goroutine happens to be running on. The main thread is of course busy running its message loop at this point (via `wde.Run`), and prior to commit 3ab9cbb there was no mechanism to run go code on the main thead.

This is not *necessarily* a problem, but I'm not familiar enough with OSX and gomacdraw to confidently claim that all objects accessed on these codepaths are in fact thread safe. In my environment I saw warnings in the console which amounted to certain functions complaining that they had been called from a thread other than the main thread. The functions weren't explicitly referenced by gomacdraw, but I was able to fix the warnings by scheduling `showWindow` to be called on the main thread.

As I said the crash is rare for me; in ~120 runs of wdetest I hit it twice. I've just updated origin/sqweek so that *all* window operations happen on the main thread, and with that build I'm up to 240 runs and no crashes. I'm interested to know how it goes for you, @ricochet2200 !

**Comment by skelterjohn** [comment]
Looks like the only thing that's guaranteed to happen in the main thread is
initMacDraw() (
https://github.com/skelterjohn/go.wde/blob/master/cocoa/framework/gomacdraw/gmd.m#L16
)

If we want to ensure all calls into the framework happen from the main
thread, I can advise on the best way to do that (hint:
http://gowithconfidence.tumblr.com/post/31431608488/muxing)

**Comment by ricochet2200** [comment]
I git pulled, then checked see that the new commits were present. Head is at 330ccd8d280e.  On the first run:

go run wdetest/wdetest.go
fatal error: unexpected signal during runtime execution
[signal 0xb code=0x1 addr=0x19 pc=0x7fff9213d79b]

runtime stack:
runtime.gothrow(0x4154fd0, 0x2a)
	/usr/local/go/src/runtime/panic.go:503 +0x8e
runtime.sigpanic()
	/usr/local/go/src/runtime/sigpanic_unix.go:14 +0x5e

goroutine 6 [syscall, locked to thread]:
runtime.cgocall_errno(0x40019d0, 0xc208019560, 0x0)
	/usr/local/go/src/runtime/cgocall.go:130 +0xf5 fp=0xc208019540 sp=0xc208019518
github.com/skelterjohn/go.wde/cocoa._Cfunc_openWindow(0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/:98 +0x4d fp=0xc208019560 sp=0xc208019540
github.com/skelterjohn/go.wde/cocoa.NewWindow(0xc8, 0xc8, 0x41dc260, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:99 +0x34 fp=0xc2080195c0 sp=0xc208019560
github.com/skelterjohn/go.wde/cocoa.func·001(0xc8, 0xc8, 0x0, 0x0, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:57 +0x69 fp=0xc2080195f8 sp=0xc2080195c0
github.com/skelterjohn/go%2ewde.NewWindow(0xc8, 0xc8, 0x0, 0x0, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wde.go:110 +0x5d fp=0xc208019630 sp=0xc2080195f8
main.func·002()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:45 +0x8d fp=0xc2080197e0 sp=0xc208019630
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2232 +0x1 fp=0xc2080197e8 sp=0xc2080197e0
created by main.wdetest
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:151 +0xf2

goroutine 1 [syscall, locked to thread]:
github.com/skelterjohn/go.wde/cocoa._Cfunc_NSAppRun()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/:32 +0x45
github.com/skelterjohn/go.wde/cocoa.Run()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:206 +0x1b
github.com/skelterjohn/go%2ewde.Run()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wde.go:87 +0x22
main.main()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:32 +0x30

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2232 +0x1

goroutine 5 [semacquire]:
sync.(*WaitGroup).Wait(0xc20801e060)
	/usr/local/go/src/sync/waitgroup.go:132 +0x169
main.wdetest()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:155 +0x128
created by main.main
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:31 +0x29

goroutine 7 [syscall, locked to thread]:
github.com/skelterjohn/go.wde/cocoa._Cfunc_openWindow(0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/:98 +0x4d
github.com/skelterjohn/go.wde/cocoa.NewWindow(0xc8, 0xc8, 0x405b5ed, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:99 +0x34
github.com/skelterjohn/go.wde/cocoa.func·001(0xc8, 0xc8, 0x0, 0x0, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/cocoa/cocoa_darwin.go:57 +0x69
github.com/skelterjohn/go%2ewde.NewWindow(0xc8, 0xc8, 0x0, 0x0, 0x0, 0x0)
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wde.go:110 +0x5d
main.func·002()
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:45 +0x8d
created by main.wdetest
	/Users/rick/.go/src/github.com/skelterjohn/go.wde/wdetest/wdetest.go:153 +0x118
fatal error: unexpected signal during runtime execution
[signal 0xb code=0x1 addr=0xb pc=0x7fff8818ced7]

runtime stack:
runtime.gothrow(0x4154fd0, 0x2a)
	/usr/local/go/src/runtime/panic.go:503 +0x8e
runtime.sigpanic()
	/usr/local/go/src/runtime/sigpanic_unix.go:14 +0x5e
exit status 2

It crashed 6/10 times.  About where it was before the commit.  AFAI can tell, all the other crash reports look very similar.  Maybe rollback the commits until we find the culprit (running 240+ times sounds painful)? Also try running with `-race`, it slows the program's execution speed.  It might make crashes happen more frequently...or less...it's worth a shot.

**Comment by sqweek** [comment]
... for some reason I left `NewWindow` out of the on-main-thread treatment, despite the fact its where you're crashing pretty consistently. I've pushed another commit -- last try on this angle I promise. Thanks for sticking with me so far!

**Comment by ricochet2200** [comment]
30 runs and no crashes, this is definitely better!

Go 1.5 just came out so I upgraded and tested another 10 runs.  They also all ran perfectly.  As far as I can tell, this branch works great on OS X.

**Comment by sqweek** [comment]
Hallelujah! I'm off to sleep now but I'll get this cleaned up and pushed to master over the next few days. :)

**Comment by skelterjohn** [comment]
Thanks for the great work, you two.

**Pull request merged**